File: sentimental-credit/credit.py
# TODO

from cs50 import get_string
import re

# American Express uses 15-digit numbers
# amex_digits = 15;
# MasterCard uses 16-digit numbers,
# master_digits = 16;
# Visa uses 13- and 16-digit numbers
visa_digits = [13, 16]

# amex_start1 = 34;
# amex_start2 = 37;
# master_start1 = 51;
# master_start2 = 52;
# master_start3 = 53;
# master_start4 = 54;
# master_start5 = 55;
# visa_start = 4;

def check_card(card_number):
    # sum = 0
    # i = 0
    # while (int(card_number) > 0):
    #     rem = int(card_number) % 10
    #     card_number = (int(card_number) - rem) / 10
    #     if i % 2 == 0:
    #         # print(rem)
    #         sum = sum + rem
    #     else:
    #         doubled = rem * 2
    #         if doubled > 9:
    #             doubled = 1 + (doubled - 10)
    #         sum = sum + doubled
    #     i += 1
    #     # print(sum)

    sum = 0
    reverse_number = card_number[::-1]
    for i in range(len(card_number)):
        if len(reverse_number) % 2 == 0:
            sum = sum + int(reverse_number[i])
        else:
            doubled = reverse_number[i] * 2
            if doubled > 9:
                doubled = 1 + (doubled - 10)
            sum = sum + doubled

        print(sum)
    return sum

def is_valid(sum):
    if sum % 10 == 0:
        return True
    return False

while True:
    card_number = get_string("Number: ")
    if int(card_number) > 0:
        break

sum = check_card(card_number)

pattern_amex = re.search("^34|^37", card_number)
pattern_master = re.search("^51|^52|^53|^54|^55", card_number)
pattern_visa = re.search("^4", card_number)

if pattern_amex and len(card_number) == 15 and is_valid(sum):
    print("AMEX")
elif pattern_master and len(card_number) == 16 and is_valid(sum):
    print("MASTERCARD")
elif pattern_visa and len(card_number) in visa_digits and is_valid(sum):
    print("VISA")
else:
    print("INVALID")

File: CS50x/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <stdbool.h>

long get_number(void);
int calc_sum(long number);
bool is_valid(int sum);
string check_type(int sum, long number);

int main(void)
{
    // Ask for a credit card number
    long number = get_number();

    // Calc the sum of the credit card number's digits
    int sum = calc_sum(number);

    // Check the type of the credit card
    string type = check_type(sum, number);
    printf("%s\n", type);
}

long get_number(void)
{
    long number;
    do
    {
        number = get_long("Number: ");
    }
    while (number < 0);
    return number;
}

int calc_sum(long number)
{
    int sum = 0;
    // The loop goes until number = 0
    for (int i = 0; number > 0; i++)
    {
        int rem = number % 10;
        number = (number - rem) / 10;
        if (i % 2 == 0)
        {
            sum = sum + rem;
        }
        else
        {
            int doubled = rem * 2;
            if (doubled > 9)
            {
                doubled = 1 + (doubled - 10);
            }
            sum = sum + doubled;
        }
    }
    return sum;
}

bool is_valid(int sum)
{
    if (sum % 10 == 0)
    {
        return true;
    }
    return false;
}

string check_type(int sum, long number)
{
    // American Express uses 15-digit numbers
    int amex_digits = 15;
    // MasterCard uses 16-digit numbers,
    int master_digits = 16;
    // Visa uses 13- and 16-digit numbers
    int visa_digits1 = 13;
    int visa_digits2 = 16;

    int amex_start1 = 34;
    int amex_start2 = 37;
    int master_start1 = 51;
    int master_start2 = 52;
    int master_start3 = 53;
    int master_start4 = 54;
    int master_start5 = 55;
    int visa_start = 4;

    // Check if the credit card number is valid
    if (is_valid(sum) == false)
    {
        return "INVALID";
    }

    for (int i = 0; i < master_digits; i++)
    {
        int rem = number % 10;
        number = (number - rem) / 10;
        // American Express numbers start with 34 or 37
        if ((number == amex_start1 && i == (amex_digits - 3)) || (number == amex_start2 && i == (amex_digits - 3)))
        {
            return "AMEX";
        }
        // MasterCard numbers start with 51, 52, 53, 54, or 55
        else if ((number == master_start1 && i == (master_digits - 3)) || (number == master_start2 && i == (master_digits - 3))
                 || (number == master_start3 && i == (master_digits - 3)) || (number == master_start4 && i == (master_digits - 3))
                 || (number == master_start5 && i == (master_digits - 3)))
        {
            return "MASTERCARD";
        }
        // Visa numbers start with 4
        else if ((number == visa_start && i == (visa_digits1 - 2)) || (number == visa_start && i == (visa_digits2 - 2)))
        {
            return "VISA";
        }
    }
    return "INVALID";
}
